packages/openapi-v3/src/extension-point/spec-contributor.service.ts

```diff
@@ -0,0 +1,69 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/example-greeter-extension
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {config, Getter} from '@loopback/context';
+import {extensionPoint, extensions} from '@loopback/core';
+import * as debugModule from 'debug';
+import * as _ from 'lodash';
+import {inspect} from 'util';
+import {OpenApiSpec} from '../types';
+import {OAISpecEnhancer, OAISPEC_ENHANCER_EXTENSION_POINT_NAME} from './types';
+const debug = debugModule('loopback:openapi:spec-contributor');
+
+/**
+ * Options for the OAI Spec contributor extension point
+ */
+export interface OAISpecContributorServiceOptions {
+  // no-op
+}
+
+/**
+ * An extension point for OAI contributors
+ */
+@extensionPoint(OAISPEC_ENHANCER_EXTENSION_POINT_NAME)
+export class OAISpecEnhancerService {
+  constructor(
+    /**
+     * Inject a getter function to fetch spec contributors
+     */
+    @extensions()
+    private getContributors: Getter<OAISpecEnhancer[]>,
+    /**
+     * An extension point should be able to receive its options via dependency
+     * injection.
+     */
+    @config()
+    public readonly options?: OAISpecContributorServiceOptions,
+  ) {}
+
+  private _spec: OpenApiSpec = {
+    openapi: '3.0.0',
+    info: {
+      title: 'LoopBack Application',
+      version: '1.0.0',
+    },
+    paths: {},
+  };
+
+  get spec(): OpenApiSpec {
+    return this._spec;
+  }
+  set spec(value: OpenApiSpec) {
+    this._spec = value;
+  }
+
+  /**
+   * Generate OpenAPI spec from contributors
+   */
+  async generateSpec(options = {}): Promise<OpenApiSpec> {
+    const contributors = await this.getContributors();
```

Comment:
How are we going to decide in which order the contributors are applied? What if we have two contributors adding conflicting values for the same field in the spec - how is the user going to determine which contributor wins?

Reply:
Good question * 2, I share your concern when implementing the extension point but haven't figured out a concrete plan, maybe use a combination of  **group** and **alphabet order**? like how we load observers group by group in https://loopback.io/doc/en/lb4/Life-cycle.html#observer-groups

Reply:
Moreover, if extension only contributes spec, and the app decides ALL the merge logic, then the order won't be a problem...

So the design here is related to our discussion in https://github.com/strongloop/loopback-next/pull/4258#discussion_r355691708.

Reply:
> maybe use a combination of group and alphabet order? like how we load observers group by group in https://loopback.io/doc/en/lb4/Life-cycle.html#observer-groups

Sounds good, I like the idea of using a mechanism we already have in place elsewhere in the framework. This way our users have less to learn.

I think it's ok to leave custom order out of scope of the initial implementation, as long as the design makes it possible to introduce custom order later.